### PR TITLE
[REVIEW] Fix apply_boolean_mask issue with non-null string column

### LINKED ISSUE #3714: [BUG] apply_boolean_mask misbehaves when inputs are not nullable
**Describe the bug**
`apply_boolean_mask`  is giving the wrong results for certain small width data types and the output bitmask is filled for strings but not other data types

**Steps/Code to reproduce bug**
`template <typename T>
struct ApplyFilter : public cudf::test::BaseFixture {};

TYPED_TEST_CASE(ApplyFilter, cudf::test::NumericTypes);
TYPED_TEST(ApplyFilter, withNull)
{
    using T = TypeParam;
    cudf::test::fixed_width_column_wrapper<T> col1{{5, 4, 3, 5, 8, 5, 6}, {1, 1, 0, 1, 1, 1, 1}};
    cudf::test::strings_column_wrapper col2({"d", "e", "a", "d", "k", "d", "l"}, {1, 0, 1, 1, 1, 1, 1});
    cudf::test::fixed_width_column_wrapper<T> col3{{10, 40, 70, 5, 2, 10, 11}, {0, 1, 1, 1, 1, 1, 0}};
    cudf::table_view cudf_table_in_view {{col1, col2, col3}};

    cudf::test::fixed_width_column_wrapper<cudf::experimental::bool8> bool_filter{{1, 1, 1, 0, 1, 0, 0}, {1, 1, 0, 1, 1, 1, 0}};
    cudf::column_view bool_filter_col(bool_filter);

    std::unique_ptr<cudf::experimental::table> filteredTable = cudf::experimental::apply_boolean_mask(cudf_table_in_view,bool_filter_col);
    cudf::table_view tableView = filteredTable->view();
    
    cudf::test::fixed_width_column_wrapper<T> expect_col1{{5, 4, 8}, {1, 1,1}};
    cudf::test::strings_column_wrapper expect_col2({"d", "e", "k"}, {1, 0, 1});
    cudf::test::fixed_width_column_wrapper<T> expect_col3{{10, 40, 2}, {0, 1, 1}};
    cudf::table_view expect_cudf_table_view {{expect_col1, expect_col2, expect_col3}};

    std::string col0_string = cudf::test::to_string(tableView.column(0), "|");
    if (tableView.column(0).nullable())
        col0_string = col0_string + " is nullable";
    std::cout<<"col0_string: "<<col0_string<<std::endl;    
    std::string col1_string = cudf::test::to_string(tableView.column(1), "|");
    if (tableView.column(1).nullable())
        col1_string = col1_string + " is nullable";
    std::cout<<"col1_string: "<<col1_string<<std::endl;
    std::string col2_string = cudf::test::to_string(tableView.column(2), "|");
    if (tableView.column(2).nullable())
        col2_string = col2_string + " is nullable";
    std::cout<<"col2_string: "<<col2_string<<std::endl;

    cudf::test::expect_tables_equal(expect_cudf_table_view, tableView);
}

TYPED_TEST(ApplyFilter, withoutNull)
{
    using T = TypeParam;
    cudf::test::fixed_width_column_wrapper<T> col1{{5, 4, 3, 5, 8, 5, 6}};
    cudf::test::strings_column_wrapper col2({"d", "e", "a", "d", "k", "d", "l"});
    cudf::test::fixed_width_column_wrapper<T> col3{{10, 40, 70, 5, 2, 10, 11}};
    cudf::table_view cudf_table_in_view {{col1, col2, col3}};

    cudf::test::fixed_width_column_wrapper<cudf::experimental::bool8> bool_filter{{1, 1, 0, 0, 1, 0, 0}};
    cudf::column_view bool_filter_col(bool_filter);

    std::unique_ptr<cudf::experimental::table> filteredTable = cudf::experimental::apply_boolean_mask(cudf_table_in_view,bool_filter_col);
    cudf::table_view tableView = filteredTable->view();
    
    cudf::test::fixed_width_column_wrapper<T> expect_col1{{5, 4, 8}};
    cudf::test::strings_column_wrapper expect_col2({"d", "e", "k"});
    cudf::test::fixed_width_column_wrapper<T> expect_col3{{10, 40, 2}};
    cudf::table_view expect_cudf_table_view {{expect_col1, expect_col2, expect_col3}};

    std::string col0_string = cudf::test::to_string(tableView.column(0), "|");
    if (tableView.column(0).nullable())
        col0_string = col0_string + " is nullable";
    std::cout<<"col0_string: "<<col0_string<<std::endl;    
    std::string col1_string = cudf::test::to_string(tableView.column(1), "|");
    if (tableView.column(1).nullable())
        col1_string = col1_string + " is nullable";
    std::cout<<"col1_string: "<<col1_string<<std::endl;
    std::string col2_string = cudf::test::to_string(tableView.column(2), "|");
    if (tableView.column(2).nullable())
        col2_string = col2_string + " is nullable";
    std::cout<<"col2_string: "<<col2_string<<std::endl;

    cudf::test::expect_tables_equal(expect_cudf_table_view, tableView);
}`

**Expected behavior**
The unit test provided above should pass

**Environment overview (please complete the following information)**
Using cudf nightly from 01/06/20 installed using conda

**Environment details**
Please run and paste the output of the `cudf/print_env.sh` script here, to gather any other relevant environment details

**Additional context**
Unit test originally was run in a combination of source code and conda installed library. Bug was discovered in BlazingSQL unit test and was then distilled to only the cudf code.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Mask was getting created by default in gather which caused this issue. There are other changes to test cases of JOIN as it was dependent on gather and with `ignore_out_of_bound` being `true` for full/left join affects it.

closes #3714

### CODE REVIEW

**Comment by rgsl888prabhu** [comment]
> Are you keeping the `tests/strings/gather_tests.cu` ?

I thought it would be better to have the out of bound test cases, if you feel I can move it to gather, will move them.

File: cpp/include/cudf/strings/detail/gather.cuh [hunk 1]
```diff
@@ -73,7 +73,11 @@ std::unique_ptr<cudf::column> gather( strings_column_view const& strings,
       mstate = ALL_NULL;
     }
     // create null mask -- caller must update this
-    rmm::device_buffer null_mask = create_null_mask( output_count, mstate, stream, mr );
+
+    rmm::device_buffer null_mask{};
```
Comment (jrhemstad) [inline cpp/include/cudf/strings/detail/gather.cuh]: This looks like a good case to use an [`IIFE`](https://www.bfilipek.com/2016/11/iife-for-complex-initialization.html) (thanks to @codereport for introducing me to these).

```suggestion
    rmm::device_buffer null_mask  = [&]{
       if(strings.parent().nullable() or NullifyOutOfBounds)
          return create_null_mask(output_count, mstate, stream, mr);
       else
          return device_buffer{};
    }();
```

File: cpp/tests/copying/gather_tests.cu [hunk 2]
```diff
@@ -348,3 +349,70 @@ TEST_F(GatherTestStr, StringColumn) {
...
+        expected_validity.begin());
+    cudf::test::expect_columns_equal(results->view().column(0),expected);
+}
+
+TEST_F(GatherTestStr, GatherZeroSizeStringsColumn)
+{
+    cudf::column_view zero_size_strings_column( cudf::data_type{cudf::STRING}, 0, nullptr, nullptr, 0);
+    auto strings_view = cudf::strings_column_view(zero_size_strings_column);
+    thrust::device_vector<int32_t> gather_map;
+    auto results = cudf::strings::detail::gather<true>(strings_view, gather_map.begin(), gather_map.end() );
```
Comment (davidwendt) [inline cpp/tests/copying/gather_tests.cu]: I was expecting that the tests would call just the `gather` API and not the `strings::detail::gather` internal function. 
Is there a way to accomplish this without calling this internal function?

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Pull request merged**